pkg/billing/model.go

```diff
@@ -4,6 +4,34 @@ import (
 	"time"
 )
 
+type Event[Self any] interface {
+	AbsoluteEvent | IncrementalEvent
+
+	eventMethods[Self]
+}
+
+// eventMethods exists because Go does not let you assume that a value of generic type "Foo | Bar"
+// has the fields common to both Foo and Bar. So we still have to thread through these fields.
+//
+// The methods in this interface *could* have directly been a part of Event, but then we couldn't do
+// the "implements eventMethods" assertions below (because you aren't allowed to store a value of
+// type "Foo | Bar"; it can only be used as a constraint for generics).
+//
+// eventMethods also takes Self as a parameter because we want Event[...] to be implemented for the
+// raw AbsoluteEvent and IncrementalEvent types, not pointers to them, because that would make
+// things just a bit more annoying. But we still need to set field values, which we can ONLY do by
+// taking the address of the field *somehow*, so we use this interface to return functions to do
+// all that stuff for us.
+type eventMethods[Self any] interface {
+	typeSetter() func(*Self)
+	idempotencyKeyGetter() func(*Self) *string
+}
+
+var (
+	_ eventMethods[AbsoluteEvent]    = AbsoluteEvent{}    //nolint:exhaustruct // just checking they impl the interface
+	_ eventMethods[IncrementalEvent] = IncrementalEvent{} //nolint:exhaustruct // just checking they impl the interface
+)
```

Comment:
while this is a fine thing to have, doesn't the compiler ensure this because you construct `Event` objects with each/either of these types?

Reply:
We're not actually using `AbsoluteEvent`s unfortunately.